File: examples/batch_example.py
#!/usr/bin/env python3
import numpy as np
import time
import sys
import os

np.set_printoptions(linewidth=1e8)

# Add the build directory to the Python path
sys.path.append(os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'build'))

# Import the Python module
import pysqpcpu


def main():
    # Path to the URDF file
    urdf_filename = "/Users/emreadabag/code/indy-ros2/indy_description/urdf_files/indy7.urdf"
    
    # Parameters
    batch_size = 4
    N = 32
    dt = 0.01
    max_qp_iters = 5
    num_threads = 4  # Use 4 threads
    
    # Create a BatchThneed instance
    batch_solver = pysqpcpu.BatchThneed(
        urdf_filename=urdf_filename,
        batch_size=batch_size,
        N=N,
        dt=dt,
        max_qp_iters=max_qp_iters,
        num_threads=num_threads
    )
    
    # Create a single Thneed instance to get dimensions
    single_solver = pysqpcpu.Thneed(urdf_filename, N, dt, max_qp_iters)
    nx = single_solver.nx
    
    # Create batch inputs
    xs_batch = []
    eepos_g_batch = []
    
    # Initialize batch inputs with different values
    for i in range(batch_size):
        xs = np.zeros(nx)
        eepos_g = np.ones(3 * N)
        
        xs_batch.append(xs)
        eepos_g_batch.append(eepos_g)
    
    # Measure execution time
    start_time = time.time()
    
    # set xs for each solver
    batch_solver.batch_update_xs(xs_batch)

    # Run batch SQP
    batch_solver.batch_sqp(xs_batch, eepos_g_batch)
    
    # Calculate elapsed time
    elapsed = (time.time() - start_time) * 1000  # Convert to ms
    
    # Get results
    results = batch_solver.get_results()
    results_np = np.array(results)
    
    # Print results
    for i, result in enumerate(results):
        print(f"Result {i}: {result[:13]} ...")
    
    t = pysqpcpu.Thneed(urdf_filename, N, dt, max_qp_iters)
    t.setxs(xs_batch[0])
    t.sqp(xs_batch[0], eepos_g_batch[0])
    print(f"Truth: {t.XU[:13]}")
    
    print(f"Total execution time: {elapsed:.2f} ms")
    print(f"Average time per problem: {elapsed / batch_size:.2f} ms\n")
    
    
    ground_truth_dist = np.max(np.abs(t.XU - results_np.max(axis=0)))
    batch_dist = np.max(np.abs(results_np.max(axis=0) - results_np.min(axis=0)))
    
    print(f"Ground truth match: {ground_truth_dist < 1e-6}, {ground_truth_dist}")
    print(f"Batch consistency: {batch_dist < 1e-6}, {batch_dist}")
    
if __name__ == "__main__":
    main() 

File: examples/batch_example.cpp
#include "../include/batch_thneed.hpp"
#include <iostream>
#include <iomanip>
#include <sys/time.h>
#include <vector>

int main(int argc, char ** argv)
{
  using namespace pinocchio;
  
  const std::string urdf_filename = "../urdfs/indy7.urdf";
  
  // Parameters
  const int batch_size = 4;
  const int N = 32;
  const float dt = 0.01;
  const int max_qp_iters = 5;
  const int num_threads = 4;  // Use 4 threads
  
  // Create a BatchThneed instance
  sqpcpu::BatchThneed batch_solver(urdf_filename, "", "end_effector", batch_size, N, dt, max_qp_iters, num_threads);
  
  // Create batch inputs
  std::vector<Eigen::VectorXd> xs_batch;
  std::vector<Eigen::VectorXd> eepos_g_batch;
  
  // Get the dimensions from the first solver
  int nx = 12;  // This should match the nx from Thneed
  
  // Initialize batch inputs with different values
  Eigen::VectorXd xs = Eigen::VectorXd::Zero(nx);
  Eigen::VectorXd eepos_g = Eigen::VectorXd::Ones(3 * N);
  
  
  // Measure execution time
  struct timeval start, end;
  gettimeofday(&start, NULL);
  
  // Run batch SQP
  batch_solver.batch_sqp(xs, eepos_g);
  
  gettimeofday(&end, NULL);
  double elapsed = (end.tv_sec - start.tv_sec) * 1000.0;    // sec to ms
  elapsed += (end.tv_usec - start.tv_usec) / 1000.0;        // us to ms
  
  // Get results
  std::vector<Eigen::VectorXd> results = batch_solver.get_results();
  
  // Print results
  for (int i = 0; i < batch_size; i++) {
    std::cout << "Result " << i << ": " << results[i].transpose().head(10) << " ..." << std::endl;
  }
  
  std::cout << "Total execution time: " << elapsed << " ms" << std::endl;
  std::cout << "Average time per problem: " << elapsed / batch_size << " ms" << std::endl;
  
  return 0;
} 